Shell/AST.h

```diff
@@ -0,0 +1,730 @@
+    JobValue(RefPtr<Job> job)
+        : m_job(move(job))
+    {
+    }
+
+    const RefPtr<Job> job() const { return m_job; }
+
+private:
+    RefPtr<Job> m_job;
+};
+
+class ListValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~ListValue();
+    virtual bool is_list() const override { return true; }
+    ListValue(Vector<RefPtr<Value>> values)
+        : m_contained_values(move(values))
+    {
+    }
+
+private:
+    Vector<RefPtr<Value>> m_contained_values;
+};
+
+class StringListValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override { return m_contained_values; }
+    virtual ~StringListValue();
+    virtual bool is_list() const override { return true; }
+    StringListValue(Vector<String> values)
+        : m_contained_values(move(values))
+    {
+    }
+
+private:
+    Vector<String> m_contained_values;
+};
+
+class StringValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~StringValue();
+    virtual bool is_string() const override { return m_split.is_null(); }
+    virtual bool is_list() const override { return !m_split.is_null(); }
+    StringValue(String string, String split_by = {})
+        : m_string(string)
+        , m_split(move(split_by))
+    {
+    }
+
+private:
+    String m_string;
+    String m_split;
+};
+
+class GlobValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~GlobValue();
+    virtual bool is_glob() const override { return true; }
+    GlobValue(String glob)
+        : m_glob(glob)
+    {
+    }
+
+private:
+    String m_glob;
+};
+
+class SimpleVariableValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~SimpleVariableValue();
+    // FIXME: Should override is_list and is_string.
+    SimpleVariableValue(String name)
+        : m_name(name)
+    {
+    }
+
+private:
+    String m_name;
+};
+
+class SpecialVariableValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~SpecialVariableValue();
+    // FIXME: Should override is_list and is_string.
+    SpecialVariableValue(char name)
+        : m_name(name)
+    {
+    }
+
+private:
+    char m_name;
+};
+
+class TildeValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~TildeValue();
+    virtual bool is_string() const override { return true; }
+    TildeValue(String name)
+        : m_username(name)
+    {
+    }
+
+private:
+    String m_username;
+};
+
+class Node : public RefCounted<Node> {
+public:
+    virtual void dump(int level) = 0;
+    virtual RefPtr<Value> run(TheExecutionInputType) = 0;
+    virtual void highlight_in_editor(Line::Editor&, RefPtr<Shell>, HighlightMetadata = {}) = 0;
```

Comment:
(I haven't looked how this used yet, but) why not
```suggestion
    virtual void highlight_in_editor(Line::Editor&, Shell&, HighlightMetadata = {}) = 0;
```
?

Reply:
hmmm...I just don't like passing normal references to `Core::Object`s, it feels weird.

Reply:
No, it's completely fine (and preferred) to pass references to refcounted objects *forward* (in arguments), but not backward (in return values)